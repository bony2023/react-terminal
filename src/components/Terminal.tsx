import * as React from "react";

import { StyleContext } from "../contexts/StyleContext";
import { ThemeContext } from "../contexts/ThemeContext";
import { useClickOutsideEvent } from "../hooks/terminal";

import Controls from "./Controls";
import Editor from "./Editor";

import Utils from "../common/Utils";

interface TerminalProps {
  enableInput: boolean
  caret: boolean
  theme: string
  showControlBar: boolean
  showControlButtons: boolean
  controlButtonLabels: string[]
  prompt: string
  commands: Record<string, (...args: never) => void>
  welcomeMessage: string | (() => void) | React.ReactNode
  errorMessage: string | ((...args: never) => void) | React.ReactNode
  defaultHandler: ((...args: never) => void) | null
};

const Terminal: React.FC<TerminalProps> = ({
  enableInput = true,
  caret = true,
  theme = "light",
  showControlBar = true,
  showControlButtons = true,
  controlButtonLabels = ["close", "minimize", "maximize"],
  prompt = ">>>",
  commands = {},
  welcomeMessage = "",
  errorMessage = "not found!",
  defaultHandler = null,
}) => {
  const wrapperRef = React.useRef(null);
  const [consoleFocused, setConsoleFocused] = React.useState(!Utils.isMobile());
  const style = React.useContext(StyleContext);
  const themeStyles = React.useContext(ThemeContext);

  useClickOutsideEvent(wrapperRef, consoleFocused, setConsoleFocused);

  const controls = showControlBar ? <Controls
    consoleFocused={consoleFocused}
    showControlButtons={showControlButtons}
    controlButtonLabels={controlButtonLabels}/> : null;

  const editor = <Editor
    caret={caret}
    consoleFocused={consoleFocused}
    prompt={prompt}
    commands={commands}
    welcomeMessage={welcomeMessage}
    errorMessage={errorMessage}
    enableInput={enableInput}
    showControlBar={showControlBar}
    defaultHandler={defaultHandler}
  />

  return (
    <div
      ref={wrapperRef}
      id={style.terminalContainer}
      className={style[`theme--${theme}`]}
      data-testid="terminal"
    >
      <div className={`${style.terminal}`} style={{ background: themeStyles.themeToolbarColor, color: themeStyles.themeColor }}>
        {controls}
        {editor}
      </div>
    </div>
  );
}

export default Terminal;