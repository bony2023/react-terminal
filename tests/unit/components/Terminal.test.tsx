import * as React from "react";
import { render } from "@testing-library/react"
import ContextProvider from "../../../src/contexts";
import { TerminalContextProvider } from "../../../src/contexts/TerminalContext";
import Terminal from "../../../src/components/Terminal";

let props: any;
const renderWrapper = () => (
  <ContextProvider>
    <TerminalContextProvider>
      <Terminal {...props} />
    </TerminalContextProvider>
  </ContextProvider>
);

describe("Terminal", () => {
  it("Terminal renders correctly", () => {
    const { container } = render(renderWrapper())
    expect(container).toMatchSnapshot();
  });

  it("Terminal doesn't render control buttons", () => {
    props = {
      showControlButtons: false,
    };
    const { container } = render(renderWrapper())
    expect(container).toMatchSnapshot();
  });

  it("Terminal doesn't render control bar", () => {
    props = {
      showControlBar: false,
    };
    const { container } = render(renderWrapper())
    expect(container).toMatchSnapshot();
  });
});
