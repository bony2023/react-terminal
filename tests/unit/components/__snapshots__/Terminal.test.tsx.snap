// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal Terminal doesn't render control bar 1`] = `
<div>
  <div
    class="theme--light"
    data-testid="terminal"
    id="terminalContainer"
  >
    <div
      class="terminal"
      style="background: rgb(216, 216, 216); color: rgb(51, 51, 51);"
    >
      <div
        class="editor curvedTop null"
        id="terminalEditor"
        style="background: rgb(253, 246, 228);"
      >
        <span
          style="color: rgb(169, 23, 168);"
        >
          &gt;&gt;&gt;
        </span>
        <div
          class="lineText"
        >
          <span
            class="preWhiteSpace"
          />
          <span
            class="caret"
          >
            <span
              class="caretAfter"
              style="background: rgb(51, 51, 51);"
            />
          </span>
          <span
            class="preWhiteSpace"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Terminal Terminal doesn't render control buttons 1`] = `
<div>
  <div
    class="theme--light"
    data-testid="terminal"
    id="terminalContainer"
  >
    <div
      class="terminal"
      style="background: rgb(216, 216, 216); color: rgb(51, 51, 51);"
    >
      <div
        class="controls"
      />
      <div
        class="editor null editorWithTopBar"
        id="terminalEditor"
        style="background: rgb(253, 246, 228);"
      >
        <span
          style="color: rgb(169, 23, 168);"
        >
          &gt;&gt;&gt;
        </span>
        <div
          class="lineText"
        >
          <span
            class="preWhiteSpace"
          />
          <span
            class="caret"
          >
            <span
              class="caretAfter"
              style="background: rgb(51, 51, 51);"
            />
          </span>
          <span
            class="preWhiteSpace"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Terminal Terminal renders correctly 1`] = `
<div>
  <div
    class="theme--light"
    data-testid="terminal"
    id="terminalContainer"
  >
    <div
      class="terminal"
      style="background: rgb(216, 216, 216); color: rgb(51, 51, 51);"
    >
      <div
        class="controls"
      >
        <div
          class="consoleCtrl close"
        />
        <div
          class="consoleCtrl minimize"
        />
        <div
          class="consoleCtrl maximize"
        />
      </div>
      <div
        class="editor null editorWithTopBar"
        id="terminalEditor"
        style="background: rgb(253, 246, 228);"
      >
        <span
          style="color: rgb(169, 23, 168);"
        >
          &gt;&gt;&gt;
        </span>
        <div
          class="lineText"
        >
          <span
            class="preWhiteSpace"
          />
          <span
            class="caret"
          >
            <span
              class="caretAfter"
              style="background: rgb(51, 51, 51);"
            />
          </span>
          <span
            class="preWhiteSpace"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
